### ISSUE TITLE
transaction aborted, when multiple requests at once

### ISSUE THREAD

**Issue by gbkwiatt** [knex/knex#4023]
Knex CLI version: 0.21.1
Knex Local version: 0.21.5
mssql: 6.2.1
node: 14.0.0
This is my test  code: 
```javascript
import knex from 'knex'

const testDB = {
  client: 'mssql',
  connection: {
    host: 'localhost',
    user: process.env.DB_USER || 'username',
    password: process.env.DB_PASSWORD || 'password',
    database: process.env.DB_DBNAME_TEST || 'test',
    options: {
      enableArithAbort: true,
      encrypt: false,
    },
  },
  pool: { min: 2, max: 7 },
}
const db = knex(testDB)
const run = async () => {
  let newRequest = { priority: 'medium', status: 'NEW', categoryID: 1, title: 'test', requestID: String(Math.floor(Math.random() * 9999)), requestorID: 'test' }
  try {
    const request = await db.transaction(async (trx) => {
      await db('requests').insert(newRequest).transacting(trx)
      return await db('requests').select('requestID').first().orderBy('createdAt', 'desc').transacting(trx)
    })
    console.log(request)
  } catch (err) {
    console.log(err.message)
  }
}
run()


```

so i wanted to test knex tranasaction the hard way. Above code works when I execute it one buy one.
But when I spam it with 10 reuqests at once, only first 2 go through and rest gives error:
```console
TransactionError: Transaction has been aborted.
    at Transaction._rollback (C:\Users\gbkwiatt\Desktop\Ticketing\ticketing-api-v2\node_modules\mssql\lib\base\transaction.js:237:37)
    at Transaction._rollback (C:\Users\gbkwiatt\Desktop\Ticketing\ticketing-api-v2\node_modules\mssql\lib\tedious\transaction.js:77:11)
    at C:\Users\gbkwiatt\Desktop\Ticketing\ticketing-api-v2\node_modules\mssql\lib\base\transaction.js:221:19
    at new Promise (<anonymous>)
    at Transaction.rollback (C:\Users\gbkwiatt\Desktop\Ticketing\ticketing-api-v2\node_modules\mssql\lib\base\transaction.js:220:12)
    at Transaction_MSSQL.rollback (C:\Users\gbkwiatt\Desktop\Ticketing\ticketing-api-v2\node_modules\knex\lib\dialects\mssql\transaction.js:28:21)
    at Function.transactor.rollback (C:\Users\gbkwiatt\Desktop\Ticketing\ticketing-api-v2\node_modules\knex\lib\transaction.js:278:42)
    at C:\Users\gbkwiatt\Desktop\Ticketing\ticketing-api-v2\node_modules\knex\lib\transaction.js:194:35
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

If I remove select statement, they all go through fine.

**Comment by elhigu** [comment]
Please make that complete single file example. Now there is unnecessary parts of express stuff and your DB config is not shown. No node / driver versions told etc. The issue template and instructions are there for a reason.

**Comment by elhigu** [comment]
In contributing.md and in https://stackoverflow.com/help/minimal-reproducible-example is more info what minimal reproducible code means.

**Comment by elhigu** [comment]
Though.. could be that the database you are using does something strange with locking tables.

**Comment by gbkwiatt** [comment]
I have made update in my post. I know that when I have insert in transaction, that table is locked completely for everyone for the time of transaction. 
I did some test previously about that, made delay in transaction and tried to execute it second time and it was just waiting for transaction to end and proceeding with second one. So therefore both requests went through. I was expecting something like this again here.

**Comment by elhigu** [comment]
> I have made update in my post. 

Did you try to run that code?

**Comment by elhigu** [comment]
> I have made update in my post. I know that when I have insert in transaction, that table is locked completely for everyone for the time of transaction.

Could you check if you can recreate this behavior by taking multiple commandline connections  to your server and do in each connection:

* start transaction
* insert row to the table
* select everything from the table
* commit

Also you can try making that select statement to be just `await db('requests').transcating(trx)` to make reproduction simpler.

**Comment by gbkwiatt** [comment]
I've never actually used CLI with knex, how do I start connection in cli ?

**Comment by elhigu** [comment]
I mean normal MSSQL commandline client app that has nothing to do with node.

**Comment by gbkwiatt** [comment]
> I mean normal MSSQL commandline client app that has nothing to do with node.

oh okay, will try it now

**Comment by gbkwiatt** [comment]
I did some tests in pure mssql ( driver) 
```javascript
import sql from 'mssql'

const run = async () => {
  await sql.connect('mssql://user:pass@localhost/Ticketing_dev')
  for (let i = 0; i < 11; i++) {
    transactionFn()
  }
}
run()
const transactionFn = () => {
  console.log('Transacting: ')
  const transaction = new sql.Transaction()
  transaction.begin((err) => {
    if (err) {
      console.log(err.message)
      transaction.rollback((err) => {
        if (err) console.log(err.message)
      })
    }
    new sql.Request(transaction).query(
      `INSERT INTO [Ticketing_dev].[dbo].[alerts] ([name],[lastExecution],[interval],[createdBy],[createdAt])
    VALUES
      ('test'
      ,getdate()
      ,100
      ,'abc'
      ,getdate()
      )`,
      (err, result) => {
        // insert should fail because of invalid value
        if (err) {
          console.log(err)
          transaction.rollback((err) => {
            if (err) console.log(err.message)
          })
        } else {
          new sql.Request(transaction).query('select top 1 * from [Ticketing_dev].[dbo].[alerts] order by uiD desc', (err2, result2) => {
            if (err2) {
              console.log(err2.message)
              transaction.rollback((err) => {
                if (err) console.log(err.message)
              })
            } else {
              //console.log(result2)
              transaction.commit((err) => {
                console.log('commited: ')
                if (err) console.log(err.message)
              })
            }
          })
        }
      },
    )
  })
}
```

and that looks all fine, executes all at once, then quickly return commits, and all in database...

**Comment by gbkwiatt** [comment]
I think I got something: i put catch() on select query 
```console
    info: ErrorMessageToken {
      name: 'ERROR',
      event: 'errorMessage',
      number: 1205,
      state: 51,
      class: 13,
      message: 'Transaction (Process ID 57) was deadlocked on lock resources with another process and has been chosen as the deadlock victim. Rerun the transaction.',
      serverName: 'xxx\\LOCAL_DEV',
      procName: '',
      lineNumber: 1
    }
```

So it looks like it doesn't create new transaction but trying to reuse same one ?

**Comment by gbkwiatt** [comment]
So the issue is,  that in that loop, first are executed all inserts, locking the table to each transaction, and then all selects, waiting for other transactions to finish but they won't finish because they all wait for each other. deadlock ...

No question is, how to prevent such thing ?

**Comment by gbkwiatt** [comment]
ok I think I fixed the issue. I had to turn on Read_committed_snapshot in my database.
`ALTER DATABASE "Ticketing_dev" SET READ_COMMITTED_SNAPSHOT ON;`
But that solved only aborted transaction and dead lock. 

After more testing I just realised, which stupidly I didn't know. 
When in transaction I have insert, and then I also select from same table (with knex) it also has to be in transaction otherwie I get error ` Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call?`
Which was my original issue.

**Comment by elhigu** [comment]
> When in transaction I have insert, and then I also select from same table (with knex) it also has to be in transaction otherwie I get error Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call?
Which was my original issue.

In your example code there were in the same transaction I did check that as a first thing 🤔 

```
    const request = await db.transaction(async (trx) => {
      await db('requests').insert(newRequest).transacting(trx)
      return await db('requests').select('requestID').first().orderBy('createdAt', 'desc').transacting(trx)
    })
```

both queries has `.transacting(trx)`

**Comment by elhigu** [comment]
but that is most frequent reason for this error that one fills connection pool with transactions and then tries to run a query that is not in a transaction and resolving transaction stays waiting for that query.

**Comment by gbkwiatt** [comment]
yes but that cause dead locks in this example.  But my original problem was actually related to missing .transacting(trx), which was fine for one by one request, but when I started spamming it wasn't fine, so I started looking for a problem, came to that loop in example which caused transaction aborted error. 

Yes seems like common problem in more complex queries, to be honest I don't like that transaction is locking whole table to everyone. But nothing I can do about it, that's pure mssql issue and transaction isolation.  Which would be nice to be able to change isolation level from knex, but I've seen that issue here already.
